<!-- Display image in righthand corner -->
<TABLE ALIGN=RIGHT BORDER=0 WIDTH=265 HEIGHT=196>
  <TR><TD ALIGN=RIGHT>
    <A HREF="http://www.broadinstitute.org/science/projects/mammals-models/vertebrates-invertebrates/aplysia/aplysia-genome-sequencing-project"
    TARGET=_blank>
    <IMG SRC="http://farm4.staticflickr.com/3317/4562373680_75a475bf94_n.jpg" WIDTH=320 HEIGHT=191 ALT="Sea Gooseberry"></A>
  </TD></TR>
  <TR><TD ALIGN=RIGHT> 
    <FONT SIZE=-1><em>Pleurobrachia bachei</em><BR> 
    </FONT>
    <FONT SIZE=-2> 
      Photo courtesy of  
      <A HREF="http://www.flickr.com/photos/48946234@N08/sets/72157623829490161/" 
      TARGET=_blank>Mathew Citarella</A>.
    </FONT>
  </TD></TR>
</TABLE>

<P>
The Feb. 2012 <em>Pleurobrachia bachei</em> draft assembly 
(Moroz Laboratory version 1.1) was produced by the
<A HREF="http://www.whitney.ufl.edu/research_programs/moroz-lab.htm" TARGET=_blank>Moroz Laboratory</A>,
Whitney Laboratory for Marine Bioscience, St. Augustine, FL.

<H3>Sample position queries</H3>
<P>
A genome position can be specified by the accession number of a
sequenced genomic region, an mRNA or EST, a chromosomal coordinate
range, or keywords from the GenBank description of an mRNA. The
following list shows examples of valid position queries for the
sea hare genome.  See the
<A HREF="../goldenPath/help/hgTracksHelp.html"
    TARGET=_blank>User's Guide</A> for more information.
<P>
<TABLE border=0 CELLPADDING=0 CELLSPACING=0>
    <TR><TH ALIGN=LEFT>Request:</TH>
        <TH WIDTH=14>&nbsp;</TH>
       	<TH ALIGN=LEFT>Genome Browser Response:</TH></TR>
    <TR><TD VALIGN=Top COLSPAN=3>&nbsp;</TD></TR>
    <TR><TD VALIGN=Top>scaffold_0</TD>
        <TD WIDTH=14>&nbsp;</TD>
       	<TD VALIGN=Top>Displays all of scaffold 0</TD>
    </TR>
    <TR><TD VALIGN=Top>scaffold_0:1-1000000</TD>
        <TD WIDTH=14>&nbsp;</TD>
       	<TD VALIGN=Top>Displays first million bases of scaffold 0</TD>
    </TR>
    <TR><TD VALIGN=Top NOWRAP>scaffold_0:1000000+2000</TD>
       	<TD WIDTH=14></TD>
       	<TD VALIGN=Top>Displays a region of scaffold 0 that spans 2000 bases, starting with position 1000000 </TD>
    </TR>
</TABLE>

<HR>
<H3>Additional details</H3>
<p>
<em>Pleurobrachia bachei</em> is a simple animal that lives, often in swarms, near the surface of the coastal northeast Pacific Ocean. The body of P. bachei is nearly spherical, to about 2 cm diameter, and is soft and jellylike, of glassy transparency. The outside of this spherical body possesses eight evenly-spaced longitudinal ribs, each about 1 mm wide and tapering at each end; the ribs run nearly from pole to pole. Each of the ribs is composed of about 50 small, overlapping, colorless combs, each composed of hundreds of long, compound cilia (Horridge, 1974), which beat together, thus propelling the animal. Normally, the effective stroke of a comb plate is directed towards the aboral end, so that the ctenophore typically swims mouth first (Tamm, 1973). The combs beat sequentially, passing metachronal waves along the ribs in the opposite direction to the effective stroke, and the comb rows are separately controlled so that the animal can swim forward, or backward, and even spin (Tamm and Moss, 1985). 
</p>
<p>
<em>Pleurobrachia bachei</em> is a carnivorous predator, capturing small zooplankton prey on two pinkish, contractile, filamentous tentacles, each of which has numerous fine side branches. The tentacles exit through two small holes on opposite sides of the body closer to the aboral pole and can extend out many cm beyond the animal, or can be contracted entirely into two sheaths inside the body that angle up towards the gut (Hyman, 1940).
</p>
<P>
The <em>Pleurobrachia bachei draft genome</em> was assembled using a custom approach designed to leverage the individual strengths of three popular de novo assembly packages and strategies: Velvet, SOAPdenovo, and pseudo-454 hybrid assembly with ABySS. 156,146,497bp of genome sequence were assembled from 1,924,202,519bp of 454 data and 97,391,397,588bp of Illumina HiSeq data. 19,525 gene models are currently predicted for this genome, after removing redundant gene models produced using Augustus version 2.5.5. Version 1.1 of the genome was produced by screening and removing contamination from prokaryotic sequences from the original draft genome. 
For more information on the assembly, see the Moroz Lab
<A HREF="http://128.227.123.35:8888/genomes/genome_pages/Pleurobrachia.html"
TARGET=_blank>Pleurobrachia Genome Project</A> page.
</P>
<P>
Bulk downloads of the sequence and annotation data are available via 
the Genome Browser 
<A HREF="#">FTP
server</A> or the 
<A HREF="#">Downloads</A> 
page. 
These data have <A HREF="../goldenPath/credits.html#seahare_use">specific 
conditions for use</A>. 
</P>
<P>
The Sea Gooseberry annotation tracks were generated by the Moroz Laboratory and 
its collaborators. See the 
<A HREF="/goldenPath/credits.html#seahare_credits">Credits</A> 
page for a detailed list of the organizations and individuals who contributed 
to this release. 
</P>
